### ISSUE TITLE
P1: autodiff op coverage — conv/pool/embedding/gather; aprender cannot train a CNN today

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3151]
Parent: #3146 · Tier **P1 — breadth** · Milestone **0.69.0**

## The gap

aprender has a real tape-based autograd engine (`crates/aprender-train/src/autograd/`, with CUDA and WGPU backward paths and gradient checkpointing — the last of which Burn's README still lists as *planned*). But its differentiable op set is 21 functions and every one of them is on the transformer path:

```
add add_scaled attention attention_causal gelu layer_norm matmul matmul_compute
matmul_nt matmul_nt_compute mul relu scale softmax sum swish transpose ...
```

Absent, verified 2026-09-12: `conv1d`, `conv2d`, `max_pool`, `avg_pool`, `gather`, `scatter`. `embedding` appears only inside `wgpu_block.rs`, not as a differentiable op.

**Consequence: aprender cannot train a CNN.** It cannot train anything that is not a transformer. Burn can train anything expressible in its tensor API.

## Why P1 and not P0

It is genuinely valuable and genuinely large. It does not unlock other tickets in this epic the way the three P0 substrates do, and aprender's strategic centre of gravity (per the four-pillar mission) is LLM fine-tuning, where the transformer ops already suffice. So: important, not blocking.

## Scope

Add forward + backward to `crates/aprender-train/src/autograd/ops/`:

1. `conv1d`, `conv2d` (stride, padding, dilation, groups)
2. `max_pool2d`, `avg_pool2d`, `adaptive_avg_pool2d`
3. `embedding` as a first-class differentiable op (sparse gradient accumulation)
4. `gather` / `scatter_add` — needed for MoE routing, which `ensemble/moe.rs` and `online/moe_construction.rs` already do forward-only
5. Reductions with axis support: `mean`, `max`, `min`, `prod` with correct subgradient at ties
6. `batch_norm` (train/eval modes differ — both)

## Acceptance criteria (each must be able to FAIL)

- [ ] **Finite-difference gradcheck** for every new op: central differences at `h=1e-5` on `f64`, relative error `<= 1e-6`, over randomized shapes and parameters (stride/padding/dilation/groups combinations). This is the load-bearing criterion.
- [ ] Oracle parity vs PyTorch autograd on pinned fixtures, following the pattern already established in `crates/aprender-core/tests/beat_pytorch_autograd_grad.rs`.
- [ ] Subgradient convention at `max`/`max_pool` ties is documented AND tested (which index receives the gradient when two inputs are equal).
- [ ] `embedding` backward accumulates correctly when the same index appears multiple times in a batch — the classic silent bug. Test asserts the summed gradient, not just a non-zero one.
- [ ] **End-to-end existence proof**: a small CNN trains on an MNIST-like fixture to a loss floor asserted in the test. Without this, the ops are unexercised plumbing.
- [ ] CUDA and WGPU backward paths either implement each op or fail loudly with the op name — never silently fall back to CPU while reporting the GPU device (repo rule: prove the mechanism engaged).
- [ ] Contract YAML with the gradcheck equations as proof obligations.

**Comment by noahgift** [comment]
milestone → 0.75.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: fine-tune/train surface: CRUX declarative fine-tune/distill (#4002)

**Comment by noahgift** [comment]
milestone 0.75.0 → 0.73.0: a CORRECTION per docs/roadmaps/milestone-triage-2026-09-23.md, section Corrections (#4024); operator 2026-09-23 via cf. Reason: CNN autodiff ops: general backlog (pillar D); the issue itself says Qwen3.5 fine-tuning does not need them
